engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java

```diff
@@ -0,0 +1,508 @@
+   */
+  @GenerateUncached
+  static abstract class InvokeEqualsNode extends Node {
+    static InvokeEqualsNode getUncached() {
+      return InvokeEqualsNodeGen.getUncached();
+    }
+
+    static InvokeEqualsNode build() {
+      return InvokeEqualsNodeGen.create();
+    }
+
+    abstract boolean execute(Atom selfAtom, Atom otherAtom);
+
+    @Specialization(guards = "cachedSelfAtomCtor == selfAtom.getConstructor()")
+    boolean invokeEqualsCachedAtomCtor(Atom selfAtom, Atom otherAtom,
+        @Cached("selfAtom.getConstructor()") AtomConstructor cachedSelfAtomCtor,
+        @Cached("getEqualsMethod(cachedSelfAtomCtor)") Function equalsMethod,
+        @Cached ExecuteCallNode executeCallNode,
+        @CachedLibrary(limit = "3") InteropLibrary interop) {
+      assert atomOverridesEquals(selfAtom);
+      Object ret = executeCallNode.executeCall(
+          equalsMethod,
+          null,
+          State.create(EnsoContext.get(this)),
+          new Object[]{selfAtom, otherAtom}
+      );
+      try {
+        return interop.asBoolean(ret);
+      } catch (UnsupportedMessageException e) {
+        throw new IllegalStateException(e);
+      }
+    }
+
+    @Specialization(replaces = "invokeEqualsCachedAtomCtor")
+    boolean invokeEqualsUncached(Atom selfAtom, Atom otherAtom,
+        @Cached ExecuteCallNode executeCallNode) {
+      Function equalsMethod = getEqualsMethod(selfAtom.getConstructor());
+      Object ret = executeCallNode.executeCall(
+          equalsMethod,
+          null,
+          State.create(EnsoContext.get(this)),
+          new Object[]{selfAtom, otherAtom}
+      );
+      assert InteropLibrary.getUncached().isBoolean(ret);
+      try {
+        return InteropLibrary.getUncached().asBoolean(ret);
+      } catch (UnsupportedMessageException e) {
+        throw new IllegalStateException(e);
+      }
+    }
+
+    @TruffleBoundary
+    static Function getEqualsMethod(AtomConstructor atomConstructor) {
+      Type atomType = atomConstructor.getType();
+      Function equalsFunction = atomConstructor
+          .getDefinitionScope()
+          .getMethods()
+          .get(atomType)
+          .get("==");
+      assert equalsFunction != null;
+      return equalsFunction;
+    }
+  }
+
+  /**
+   * Returns true if the given atom overrides `==` operator.
+   */
+  @TruffleBoundary
+  private static boolean atomOverridesEquals(Atom atom) {
+    var atomType = atom.getConstructor().getType();
+    Map<String, Function> methodsOnType = atom
+        .getConstructor()
+        .getDefinitionScope()
+        .getMethods()
+        .get(atomType);
+    if (methodsOnType != null) {
+      return methodsOnType.containsKey("==");
+    } else {
+      return false;
+    }
+  }
+
+  @Fallback
+  @TruffleBoundary
+  boolean equalsGeneric(Object left, Object right,
+      @CachedLibrary(limit = "5") InteropLibrary interop) {
+    EnsoContext ctx = EnsoContext.get(interop);
+    if (isHostObject(ctx, left) && isHostObject(ctx, right)) {
```

Comment:
That feels like it could be a specialization on its own?

Reply:
Probably. I just didn't want to fiddle with `@Cached` context. Besides, this whole method is behind truffle boundary.